cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+
+(1) Cypher already has good semantics for equality within the primitive types (booleans, strings, integers, and floats) and maps.
+Furthermore, Cypher has good semantics for comparability and orderability for integers, floats, and strings, within the types.
+However working with values of different types can be difficult:
+
+  * Comparability between values of different types is often undefined, which is something we regard as a problem for certain combinations of values and types.
+  * `ORDER BY` will often fail with an error if the values passed to it have different types.
+
+(2) The underlying conceptual model is complex and sometimes inconsistent which leads to an unclear relationship between comparison operators, equality, grouping, and `ORDER BY`:
+
+  * Comparability and orderability are not aligned with each other consistently (Some types may be ordered but not compared).
+  * There are various inconsistencies around equality (and equivalence) semantics as exposed by `IN`, `=`, `DISTINCT`, and grouping.
+  The difference between equality and equivalence in Cypher today is small and subtle, and limited to testing two instances of the value `null` to each other.
+
+    ** In equality, `null = null` is `null`.
+    ** In equivalence, used by grouping and `DISTINCT`, instances of `null` are always treated as being the same value.
+    ** However, equality treats `null` values differently if they are an element of a list or a map value.
+    ** Similar rules apply for `NaN`.
+
+== Proposal
+
+=== Comparability
+
+We propose that comparability be defined between any pair of values, as specified below (Please note that `null` is interpreted as meaning "unknown" or "could be any value" in the context of comparability):
+
+* Integers are compared numerically.
+* Floats (excluding `NaN` and the Infinities) are compared numerically.
+* Arbitrary numbers (excluding `NaN` and the Infinities) are compared to each other numerically, i.e. the comparison behaves as if both numbers would have been coerced to arbitrary precision big decimals (currently outside the Cypher type system) before comparing them with each other numerically.
+* Positive infinity is of type `FLOAT`, equal to itself and larger than any other number (excluding `NaN`).
+* Negative infinity is of type `FLOAT`, equal to itself and smaller than any other number (excluding `NaN`).
+* Comparing a number to a `NaN` always yields `null`.
+* Booleans are compared such that `false < true`.
+* Strings are compared in dictionary order, i.e. characters are compared pairwise in order and characters missing in a shorter string are considered to be smaller than any other character. For example, `'a' < 'aa'`.
+* Lists are compared in dictionary order, i.e. list elements are compared pairwise in order and elements missing in a shorter list are considered to be smaller than any other value. For example, `[1] < [1,0]`.
```

Comment:
Stylistically, could "List" not have sub bullets, so as to draw all the points closer together (than having the info spread out across 2 bullets, as it is at present)?